### DISCUSSION TITLE
Unit Test Assertion for Counter Fails

### DISCUSSION THREAD

**Discussion by nzxwang** [vectordotdev/vector#19594] [Q&A]
Hi,

I'm trying to write a unit test for an `aggregate` that aggregates `counter` metrics. My assertions are as follows:
```toml
    [[tests.outputs.conditions]]
    type = "vrl"
    source = '''
    assert_eq!(.tags.service, "product_1.component_1")
    assert_eq!(.tags.location, "environment_1.datacenter_1.shard_1")
    assert_eq!(.counter.value, 2)
    '''
```

The first 2 assertions on the tag work correctly, but the third assertion fails with a blatantly contradicting error message (the output payload explicitly shows the value of `.counter.value` being equal to `2.0`:
```
Running tests
test logs_usage_audit ... failed

failures:

test logs_usage_audit:

check[0] for transforms ["minutely_logs_usage_audit"] failed conditions:

  condition[0]: source execution failed: 
error[E000]: function call error for "assert_eq" at (128:151): assertion failed: null == 2
  ┌─ :3:5
  │
3 │     assert_eq!(.counter.value, 2)
  │     ^^^^^^^^^^^^^^^^^^^^^^^ assertion failed: null == 2
  │
  = see language documentation at https://vrl.dev
  = try your code in the VRL REPL, learn more at https://vrl.dev/examples


output payloads from ["minutely_logs_usage_audit"] (events encoded as JSON):
  {"name":"logs_usage_audit","tags":{"location":"environment_1.datacenter_1.shard_1","service":"product_1.component_1"},"timestamp":"2024-01-11T18:14:47.209476506Z","kind":"incremental","counter":{"value":2.0}}
```

This also fails and claims a value of `null` when I try `assert_eq!(.counter, 2)`. It seems like what I'm doing is very similar to https://github.com/vectordotdev/vector/blob/2448a72770444e4c203d7d937e1ccede22c23aed/tests/behavior/transforms/task_transform.toml#L90. Any insights appreciated!

**Comment by jszwedko** [comment]
I think you are hitting this issue: https://github.com/vectordotdev/vector/issues/5521

**Reply by nzxwang** [reply]
Thanks for the quick reply! I read through the issue and I'm having trouble understanding how remap is related to the assertion function in vrl (maybe its the same)? You did confirm it here though: https://github.com/vectordotdev/vector/issues/5521#issuecomment-1776085631

In any case, do you have a recommendation for a workaround or next best way to unit test an aggregation?

**Reply by nzxwang** [reply]
And if so, why does this work? https://github.com/vectordotdev/vector/blob/2448a72770444e4c203d7d937e1ccede22c23aed/tests/behavior/transforms/task_transform.toml#L90

**Reply by jszwedko** [reply]
Ah, interesting. Did you try that in your test? `.counter` instead of `.counter.value`?

**Reply by nzxwang** [reply]
Yes I did. That fails with the same error message claiming that the value is null.
